### PR TITLE
Turn 2 type flags into properties

### PR DESCRIPTION
**Pull request by sandersn**
This is needed for object spread and rest, which will each need a type
flag.

There are 4-5 other likely targets for removal, and I may remove those
later.

### CODE REVIEW

**Comment by sandersn** [comment]
@ahejlsberg and @mhegazy can you take a look at this?

File: src/compiler/types.ts [hunk 1]
```diff
@@ -2382,19 +2382,16 @@ namespace ts {
         Union                   = 1 << 19,  // Union (T | U)
         Intersection            = 1 << 20,  // Intersection (T & U)
         Anonymous               = 1 << 21,  // Anonymous
-        Instantiated            = 1 << 22,  // Instantiated anonymous type
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: I would preserve this flag unless you really need to free it up. In generic code we create *lots* of instantiated anonymous types and it would be nice not to incur the overhead of another property.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3177,9 +3177,7 @@ namespace ts {
             if (includePatternInType) {
                 result.pattern = pattern;
             }
-            if (hasComputedProperties) {
-                result.flags |= TypeFlags.ObjectLiteralPatternWithComputedProperties;
-            }
+            result.inObjectLiteralPatternWithComputedProperties = hasComputedProperties;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Only create this property if `hasComputedProperties` is true.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -4414,7 +4413,7 @@ namespace ts {
          * boolean, and symbol primitive types, return the corresponding object types. Otherwise return the
          * type itself. Note that the apparent type of a union type is the union type itself.
          */
-        function getApparentType(type: Type): Type {
+        function getApparentType(type: Type): ObjectType {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why this change? This function isn't guaranteed to return an `ObjectType`.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -6646,7 +6647,7 @@ namespace ts {
             }
 
             function hasExcessProperties(source: FreshObjectLiteralType, target: Type, reportErrors: boolean): boolean {
-                if (!(target.flags & TypeFlags.ObjectLiteralPatternWithComputedProperties) && maybeTypeOfKind(target, TypeFlags.ObjectType)) {
+                if (maybeTypeOfKind(target, TypeFlags.ObjectType) && !(target as ObjectType).inObjectLiteralPatternWithComputedProperties) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The `maybeTypeOfKind` function could return true when `target` is a union or intersection containing an object type, so the second check needs to first ensure `target` is actually an object type.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -9937,7 +9938,7 @@ namespace ts {
 
         // Return the contextual type for a given expression node. During overload resolution, a contextual type may temporarily
         // be "pushed" onto a node using the contextualType property.
-        function getApparentTypeOfContextualType(node: Expression): Type {
+        function getApparentTypeOfContextualType(node: Expression): ObjectType {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The returned type could be something other than an `ObjectType` so this change isn't correct.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -10371,7 +10372,8 @@ namespace ts {
             const numberIndexInfo = hasComputedNumberProperty ? getObjectLiteralIndexInfo(node, propertiesArray, IndexKind.Number) : undefined;
             const result = createAnonymousType(node.symbol, propertiesTable, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo);
             const freshObjectLiteralFlag = compilerOptions.suppressExcessPropertyErrors ? 0 : TypeFlags.FreshLiteral;
-            result.flags |= TypeFlags.ObjectLiteral | TypeFlags.ContainsObjectLiteral | freshObjectLiteralFlag | (typeFlags & TypeFlags.PropagatingFlags) | (patternWithComputedProperties ? TypeFlags.ObjectLiteralPatternWithComputedProperties : 0);
+            result.flags |= TypeFlags.ObjectLiteral | TypeFlags.ContainsObjectLiteral | freshObjectLiteralFlag | (typeFlags & TypeFlags.PropagatingFlags);
+            result.inObjectLiteralPatternWithComputedProperties = patternWithComputedProperties;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Only set the property if `patternWithComputedProperties` is true.

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [inline src/compiler/checker.ts]: Hm, turns out I was wrong when I thought that `x.prop = undefined` is a no-op. Fixed.

File: src/compiler/types.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/types.ts]: I only need one flag right now (Spread), but I'll need another for Rest and Ryan will need one for Subset. Maybe ObjectLiteral is a better target? 

FreshLiteral seems pretty easy to excise as well, but there are a lot of literals in any given program.

File: src/compiler/checker.ts (hunk 6 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 3 above)
Comment (sandersn) [inline src/compiler/checker.ts]: All the branches do return `ObjectType`, but I forgot about the else branch. Reverted.

File: src/compiler/checker.ts (hunk 4 above)
Comment (sandersn) [inline src/compiler/checker.ts]: I was relying on unions and intersections not having `inObjectLiteralPatternWithComputedProperties` and having the check come back with undefined. Adding the check is safer, though.

File: src/compiler/checker.ts (hunk 5 above)
Comment (sandersn) [inline src/compiler/checker.ts]: Fixed

File: src/compiler/types.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/types.ts]: Instantiated is back in. I'm looking at ObjectLiteral instead.

Comment (ahejlsberg) [inline src/compiler/types.ts]: I'd leave ObjectLiteral alone. Just free up the other two for now and then we can do more if and when it is needed. I'd prefer not incurring overhead unless we have to.

**Comment by sandersn** [comment]
I put `Instantiated` back in and removed `ObjectLiteral` instead. This keeps 3 flags open for use with Spread, Rest and Subset.

**Comment by sandersn** [comment]
Oops, I didn't see your new comments, @ahejlsberg. I put `ObjectLiteral` back as well.

**Comment by Igorbek** [comment]
Why didn't you just introduce a new `flags2` ?

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -3178,7 +3178,7 @@ namespace ts {
                 result.pattern = pattern;
             }
             if (hasComputedProperties) {
-                result.flags |= TypeFlags.ObjectLiteralPatternWithComputedProperties;
+                result.inObjectLiteralPatternWithComputedProperties = hasComputedProperties;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Just say `= true;`

Comment (ahejlsberg) [inline src/compiler/checker.ts]: Also, shouldn't it be `isObjectBlaBlaBla...`? In other words, "is" instead of "in".

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -10271,7 +10274,7 @@ namespace ts {
             const contextualTypeHasPattern = contextualType && contextualType.pattern &&
                 (contextualType.pattern.kind === SyntaxKind.ObjectBindingPattern || contextualType.pattern.kind === SyntaxKind.ObjectLiteralExpression);
             let typeFlags: TypeFlags = 0;
-            let patternWithComputedProperties = false;
+            let patternWithComputedProperties: boolean | undefined;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: No need for this change.

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -3155,9 +3155,9 @@ namespace ts {
         }
 
         // Return the type implied by an object binding pattern
-        function getTypeFromObjectBindingPattern(pattern: ObjectBindingPattern, includePatternInType: boolean, reportErrors: boolean): Type {
+        function getTypeFromObjectBindingPattern(pattern: ObjectBindingPattern, includePatternInType: boolean, reportErrors: boolean): ResolvedType {
             const members = createMap<Symbol>();
-            let hasComputedProperties = false;
+            let hasComputedProperties: boolean;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: No need for this change.

Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 8 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/types.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/types.ts]: done

File: src/compiler/checker.ts (hunk 7 above)
Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -3156,7 +3156,7 @@ namespace ts {
         }
 
         // Return the type implied by an object binding pattern
-        function getTypeFromObjectBindingPattern(pattern: ObjectBindingPattern, includePatternInType: boolean, reportErrors: boolean): Type {
+        function getTypeFromObjectBindingPattern(pattern: ObjectBindingPattern, includePatternInType: boolean, reportErrors: boolean): ResolvedType {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: One more change that isn't necessary.

Comment (sandersn) [inline src/compiler/checker.ts]: reverted to Type
